core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java

```diff
@@ -2784,52 +2787,66 @@ protected String preparePassword(String password, String saltValue) throws UserS
         try {
             credentialObj = Secret.getSecret(password);
         } catch (UnsupportedSecretTypeException e) {
             throw new UserStoreException("Unsupported credential type", e);
         }
+        String digestFunction = realmConfig.getUserStoreProperties().get(JDBCRealmConstants.DIGEST_FUNCTION);
+        if (digestFunction != null) {
 
-        try {
-            String passwordString;
-            if (saltValue != null) {
-                credentialObj.addChars(saltValue.toCharArray());
-            }
+            if (digestFunction.equals(UserCoreConstants.RealmConfig.PASSWORD_HASH_METHOD_PLAIN_TEXT)) {
+                passwordHash = new String(credentialObj.getChars());
+                return passwordHash;
+            } else if (digestFunction.equals(UserCoreConstants.HashConstants.PASSWORD_HASH_METHOD_SHA256)) {
+                try {
+                    if (saltValue != null) {
+                        credentialObj.addChars(saltValue.toCharArray());
+                    }
+                    MessageDigest digest = MessageDigest.getInstance(digestFunction);
+                    byte[] byteValue = digest.digest(credentialObj.getBytes());
+                    passwordHash = Base64.encode(byteValue);
 
-            String digestFunction = realmConfig.getUserStoreProperties().get(JDBCRealmConstants.DIGEST_FUNCTION);
-            if (digestFunction != null) {
-                if (digestFunction.equals(UserCoreConstants.RealmConfig.PASSWORD_HASH_METHOD_PLAIN_TEXT)) {
-                    passwordString = new String(credentialObj.getChars());
-                    return passwordString;
+                } catch (NoSuchAlgorithmException e) {
+                    String msg = "Error occurred while preparing password.";
+                    if (log.isDebugEnabled()) {
+                        log.debug(msg, e);
+                    }
+                    throw new UserStoreException(msg, e);
+                }
+            } else if (digestFunction.equals(UserCoreConstants.HashConstants.PASSWORD_HASH_METHOD_PBKDF2)) {
+                String passwordString = String.valueOf(credentialObj.getChars());
+                Map<String, Object> metaProperties = new HashMap<>();
+                metaProperties.put("iterations", 10000);
+                metaProperties.put("dkLength", 256);
+                metaProperties.put("PRF", "PBKDF2WithHmacSHA256");
+                HashProvider hashProvider = UserStoreMgtDataHolder.getInstance().getHashProvider(
+                        "PBKDF2");
+                try {
+                    byte[] hashByteArray = hashProvider.getHash(passwordString, saltValue, metaProperties);
+                    passwordHash = Base64.encode(hashByteArray);
+                } catch (HashProviderException e) {
+                    String msg = "Error in pbkdf2 hashing.";
+                    if (log.isDebugEnabled()) {
+                        log.debug(msg, e);
+                    }
```

Comment:
this logic needs to be refactored.
- If there is a `HashProvider` with the name specified in `digestFunction`, we need to invoke that method with user store properties as the meta properties.
- If not then we have to support the default OOTB behavior

Reply:
The current logic is not a generalized approach. According to this if we need to add another provider, we need to change the kernel code.